File: src/Chordpro/Line.php
<?php

namespace PhpChords\Chordpro;

class Line
{
  const REMOVE_BRACKETS_REGEX = "/(\[|\])/";
  const CHORD_REGEX = "/\[(?<chord>.{1,6})\]/";
  const CHORUS_KEY_WORDS = ["{start_of_chorus}", "{end_of_chorus}"];

  private $chorus;
  private $first_line;
  private $line;
  private $parsed = [];

  public function __construct(string $line, bool $first_line = false, bool $chorus = false)
  {
    $this->line = $line;
    $this->first_line = $first_line;
    $this->chorus = $chorus;
    $this->parsed = $this->parse($line);
  }

  public function is_chorus()
  {
    return $this->chorus;
  }
  public function is_first_line()
  {
    return $this->first_line;
  }
  public function is_text_only()
  {
    return sizeof($this->parsed) == 1 && is_null(array_values($this->parsed)[0]);
  }
  public function to_array()
  {
    return $this->parsed;
  }

  ##
  # Parses the line and returns a hash with the lyric and its chord
  # @return [Hash<String, String>] {'text' => 'chord' || nil}
  private function parse(string $line)
  {

    # Split trimmed line by regex, and clean null or empty (strlen == 0)
    $parts = array_values(array_filter(preg_split(self::CHORD_REGEX, trim($line)), 'strlen'));
    preg_match_all(self::CHORD_REGEX, trim($line), $matches,);
    $chords = $matches['chord'];

    if ($line[0] != '[') array_unshift($chords, null);

    $ret_val = [];
    foreach ($parts as $index => $value) {
      $ret_val[$value] = $chords[$index];
    }

    return $ret_val;
  }

  ##
  # Retruns and array of Lines. With the chords and chorus and first lines.
  # @param [String] the song_text
  # @return [Array<Line>] the lines of a song_text
  public static function generate_all(string $song_text)
  {
    $chorus = $first_line = false;
    $ret_val = [];
    foreach (explode("\r\n", $song_text) as $line_string) {

      if (self::is_special_line($line_string, $chorus)) {
        $first_line = true;
        continue;
      }

      $ret_val[] = new Line($line_string, $first_line, $chorus);
      $first_line = false;
    }

    return $ret_val;
  }

  ##
  # If special line is passed, it yeilds the chorus status.
  # @param [String] current_line
  # @param [boolean] <out> $chorus
  # @return [boolean] is_special
  private static function is_special_line(string $line_string, bool &$chorus)
  {
    $is_start = str_contains($line_string, self::CHORUS_KEY_WORDS[0]);
    $is_end = str_contains($line_string, self::CHORUS_KEY_WORDS[1]);
    $is_keyword = $is_start || $is_end;
    $is_empty = empty(trim($line_string));

    # Only if keyword, we modify the out parameter.
    if($is_keyword) $chorus = $is_start;

    return $is_empty || $is_keyword;
  }
}

File: TO_BE_PORTED/printers/line.rb
# frozen_string_literal: true

module Chordpro
  module Printers
    ##
    # Representation of a chunk
    class Line
      REMOVE_BRACKETS_REGEX = /(\[|\])/
      CHORD_REGEX = /\[.{1,6}\]/
      CHORUS_KEY_WORDS = %w[{start_of_chorus} {end_of_chorus}].freeze

      attr_reader :chord, :text

      define_method(:chorus?) { @chorus }
      define_method(:text_only?) { @parsed.keys.length == 1 }
      define_method(:to_h) { @parsed }

      def initialize(line, chorus: false)
        @chorus = chorus
        @line = line
        @parsed = parse(line)
      end

      private

      ##
      # Parses the line and returns a hash with the lyric and its chord
      # @return [Hash<String, String?>] {'text' => 'chord' || nil}
      def parse(line)
        parts = line.strip.split(CHORD_REGEX).delete_if(&:blank?) # ['vine a ala','bar a', 'dios']
        chords = line.strip.scan(CHORD_REGEX) # [do, re, sol]
        chords.unshift(nil) unless line[0] == '[' # [nil, do, re, sol]

        # { 'vine a ala' => 'do', 'bar a' => 'sol', 'Dios' => 'lam'} -> 'text' => 'chord' || nil
        parts.each_with_index.to_h { |part, index| [index, [part, chords[index]&.gsub(REMOVE_BRACKETS_REGEX, '')]] }
      end

      class << self
        ##
        # Retruns and array of Lines. With the chords and chorus and first lines.
        # @param [String] the song_text
        # @return [Array<Line>] the lines of a song_text
        def generate_all(song_text)
          chorus = false
          song_text.split("\r\n").map do |line|
            next if line.empty?
            next Line.new('', chorus:) if special_line?(line) { |c| chorus = c }

            # Block runs at the end of initialization (to set the variable to false in the same line)
            Line.new(line, chorus:)
          end.delete_if(&:blank?)
        end

        private

        ##
        # If special line is passed, it yeilds the chorus status.
        # @param [String] current_line
        # @return [boolean] special? if &block not given
        # @yield [boolean] chorus if &block given and is special?
        def special_line?(line)
          yield (line.include? 'start') if block_given? && CHORUS_KEY_WORDS.include?(line.strip)

          CHORUS_KEY_WORDS.include?(line.strip)
        end
      end
    end
  end
end
